include/envoy/config/incremental_subscription.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+template <class ResourceType> class IncrementalSubscriptionCallbacks {
+public:
+  typedef Protobuf::RepeatedPtrField<ResourceType> ResourceVector;
+
+  virtual ~IncrementalSubscriptionCallbacks() {}
+
+  /**
+   * Called when an incremental configuration update is received.
+   * @param added_resources resources newly added since the previous fetch.
+   * @param removed_resources names of resources that this fetch instructed to be removed.
+   * @param version_info update version.
+   * @throw EnvoyException with reason if the incremental config changes are rejected. Otherwise the
+   *        changes are accepted. Accepted changes have their version_info reflected in subsequent
+   *        requests.
+   */
+  virtual void
+  onIncrementalConfig(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                      const std::string& version_info) PURE;
+
+  /**
+   * Called when either the IncrementalSubscription is unable to fetch a config update or when
+   * onIncrementalConfig invokes an exception.
+   * @param e supplies any exception data on why the fetch failed. May be nullptr.
+   */
+  virtual void onIncrementalConfigFailed(const EnvoyException* e) PURE;
+
+  /**
+   * Obtain the "name" of a v2 API resource in a google.protobuf.Any, e.g. the route config name for
+   * a RouteConfiguration, based on the underlying resource type.
+   */
+  virtual std::string resourceName(const ProtobufWkt::Any& resource) PURE;
+};
+
+/**
+ * Common abstraction for subscribing to incremental config updates. Due to the incremental
+ * approach's stateful nature, the service can only be implemented as a gRPC bidi stream.
```

Comment:
It should be possible to eventually implement this for REST, I think. What it would require is we treat each REST request the same way we do gRPC connection resumption, and supply the full set of resource versions on each request. This might end up being useful. If you agree with this assessment, I would avoid baking in any gRPC dependencies, implement for gRPC for now, and then add some TODOs/issues to track a future REST implementation.

Reply:
I see, so you get around the statefulness by just including the necessary state, and that's fine because it's not that large - in the case of huge amounts of resources, the Envoy doesn't care about most of them, and so they can be left out entirely from both directions. That makes sense. I agree with keeping it open.

Reply:
Yeah, I don't think REST will scale particularly well, but for the cases it covers today, and in the interest of supporting on-demand via REST, we could do this.

I think for this PR, gRPC is fine, we just should leave open the path to REST implementation in the future.

Reply:
This is getting into slightly weird territory, but I just realized you could actually do it with REST without sending O(resources) versions every time, if you used a Bloom filter for the version strings the client sends!